### PR TITLE
Moved the puck configuration to actions into the vive controller plugin

### PR DESCRIPTION
**Pull request by druiz17**
Moved the puck configuration to actions into the vive controller plugin

### CODE REVIEW

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 1]
```diff
@@ -185,12 +204,113 @@ void ViveControllerManager::InputDevice::handleTrackedObject(uint32_t deviceInde
...
+                } else if (puckCount >= 4) {
+                    _config = Config::FeetHipsAndChest;
+                } else {
+                    return;
+                }
+
+                std::sort(_validTrackedObjects.begin(), _validTrackedObjects.end(), sortPucksYPosition);
+
+                auto firstFoot = _validTrackedObjects[0];
+                auto secondFoot = _validTrackedObjects[1];
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: If you made these auto& you would avoid a copy of a `std::pair<uint32_t, controller::Pose>`

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 2]
```diff
@@ -185,12 +204,113 @@ void ViveControllerManager::InputDevice::handleTrackedObject(uint32_t deviceInde
...
+                controller::Pose firstFootPose = firstFoot.second;
+                controller::Pose secondFootPose = secondFoot.second;
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: If you made these controllerPose& you would avoid a copy of a controller::Pose structure

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 3]
```diff
@@ -41,9 +45,22 @@ static const char* MENU_PARENT = "Avatar";
 static const char* MENU_NAME = "Vive Controllers";
 static const char* MENU_PATH = "Avatar" ">" "Vive Controllers";
 static const char* RENDER_CONTROLLERS = "Render Hand Controllers";
+const quint64 CALIBRATION_TIMELAPSE = 3000000;
 
 const char* ViveControllerManager::NAME { "OpenVR" };
 
+glm::mat4 computeOffset(glm::mat4 defaultToReferenceMat, glm::mat4 defaultJointMat, controller::Pose puckPose) {
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: Minor, but mark this helper function as static.  This helps prevent inadvertent linker errors, in this case it's not a big deal, but it's good practice.

    static glm::mat4 computeOffset(glm::mat4 defaultToReferenceMat, glm::mat4 defaultJointMat, controller::Pose puckPose) {

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 4]
```diff
@@ -41,9 +45,22 @@ static const char* MENU_PARENT = "Avatar";
...
+    glm::mat4 poseMat = createMatFromQuatAndPos(puckPose.rotation, puckPose.translation);
+    glm::mat4 referenceJointMat = defaultToReferenceMat * defaultJointMat;
+    return glm::inverse(poseMat) * referenceJointMat;
+}
+
+bool sortPucksYPosition(std::pair<uint32_t, controller::Pose> firstPuck, std::pair<uint32_t, controller::Pose> secondPuck) {
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: same here, mark this as static

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 5]
```diff
@@ -185,12 +204,113 @@ void ViveControllerManager::InputDevice::handleTrackedObject(uint32_t deviceInde
...
+                _calibrated = false;
+            }
+        }
+    } else {
+        _triggersPressedHandled = false;
+    }
+}
+
+void ViveControllerManager::InputDevice::updateCalibratedLimbs() {
+     _poseStateMap[controller::LEFT_FOOT] = addOffsetToPuckPose(controller::LEFT_FOOT);
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: bad ident

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 6]
```diff
@@ -185,12 +204,113 @@ void ViveControllerManager::InputDevice::handleTrackedObject(uint32_t deviceInde
...
+}
+
+controller::Pose ViveControllerManager::InputDevice::addOffsetToPuckPose(int joint) {
+    auto puck = _jointToPuckMap.find(joint);
+    if (puck != _jointToPuckMap.end()) {
+        uint32_t puckIndex = puck->second;
+        controller::Pose puckPose = _poseStateMap[puckIndex];
+        glm::mat4 puckOffset = _pucksOffset[puckIndex];
+        controller::Pose newPose = puckPose.postTransform(puckOffset);
+        return newPose;
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: I would just return the pose directly here:

    return puckPose.postTransform(puckOffset);

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 7]
```diff
@@ -185,12 +204,113 @@ void ViveControllerManager::InputDevice::handleTrackedObject(uint32_t deviceInde
...
+}
+
+void ViveControllerManager::InputDevice::updateCalibratedLimbs() {
+     _poseStateMap[controller::LEFT_FOOT] = addOffsetToPuckPose(controller::LEFT_FOOT);
+    _poseStateMap[controller::RIGHT_FOOT] = addOffsetToPuckPose(controller::RIGHT_FOOT);
+    _poseStateMap[controller::HIPS] = addOffsetToPuckPose(controller::HIPS);
+    _poseStateMap[controller::SPINE2] = addOffsetToPuckPose(controller::SPINE2);
+}
+
+controller::Pose ViveControllerManager::InputDevice::addOffsetToPuckPose(int joint) {
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: this guy doesn't mutate any internal state of the input device, so this method should be const

**Comment by hifi-gustavo** [comment]
Build console output for commit: c3d57459d5f9bfeb9e4d518446bec08b0c586fd1 available at: https://gist.github.com/9dcf8524cf932ab850f2977c07ba5be8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10412-c3d5745.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10412-c3d5745.dmg

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 8]
```diff
@@ -41,9 +45,22 @@ static const char* MENU_PARENT = "Avatar";
 static const char* MENU_NAME = "Vive Controllers";
 static const char* MENU_PATH = "Avatar" ">" "Vive Controllers";
 static const char* RENDER_CONTROLLERS = "Render Hand Controllers";
+const quint64 CALIBRATION_TIMELAPSE = 3000000;
```
Comment (sethalves) [inline plugins/openvr/src/ViveControllerManager.cpp]: or make use of USECS_PER_SECOND from libraries/shared/src/NumericalConstants.h

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 9]
```diff
@@ -185,12 +204,113 @@ void ViveControllerManager::InputDevice::handleTrackedObject(uint32_t deviceInde
...
+                // calculate the offset from the centerOfEye to defaultHeadMat
+                glm::mat4 defaultHeadOffset = glm::inverse(inputCalibration.defaultCenterEyeMat) * inputCalibration.defaultHeadMat;
+
+                currentHead = currentHead * defaultHeadOffset;
+
+                // calculate the defaultToRefrenceXform
+                glm::mat4 defaultReferenceXform = currentHead * glm::inverse(inputCalibration.defaultHeadMat);
+
+                auto puckCount = _validTrackedObjects.size();
+                if (puckCount == 2) {
```
Comment (ZappoMan) [inline plugins/openvr/src/ViveControllerManager.cpp]: coding standard: no magic numbers. This should be `MIN_PUCK_COUNT`, `MIN_FEET_AND_HIPS`, `MIN_FEET_HIPS_CHEST` or similar

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 10]
```diff
@@ -185,12 +204,113 @@ void ViveControllerManager::InputDevice::handleTrackedObject(uint32_t deviceInde
...
+                    _config = Config::FeetAndHips;
+                } else if (puckCount >= 4) {
+                    _config = Config::FeetHipsAndChest;
+                } else {
+                    return;
+                }
+
+                std::sort(_validTrackedObjects.begin(), _validTrackedObjects.end(), sortPucksYPosition);
+
+                auto firstFoot = _validTrackedObjects[0];
```
Comment (ZappoMan) [inline plugins/openvr/src/ViveControllerManager.cpp]: coding standard: no magic numbers, these indices should be something like `FIRST_FOOT_INDEX`, `SECOND_FOOT_INDEX` etc.. basically your logic appears to assume that these `_validTrackedObjects` result in a vector sorted in a particular way, so you should document that with named constants

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 11]
```diff
@@ -211,102 +220,110 @@ void ViveControllerManager::InputDevice::handleTrackedObject(uint32_t deviceInde
...
+        _config = Config::FeetHipsAndChest;
     } else {
-        _triggersPressedHandled = false;
+        return;
+    }
+
+    std::sort(_validTrackedObjects.begin(), _validTrackedObjects.end(), sortPucksYPosition);
+
+    int firstFootIndex = 0;
+    int secondFootIndex = 1;
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: make these const and capital, please

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 12]
```diff
@@ -211,102 +220,110 @@ void ViveControllerManager::InputDevice::handleTrackedObject(uint32_t deviceInde
...
+    } else {
+        uncalibrate();
+    }
+}
+
+bool ViveControllerManager::InputDevice::checkForCalibrationEvent() {
+    auto& leftTrigger = _buttonPressedMap.find(controller::LT);
+    auto& rightTrigger = _buttonPressedMap.find(controller::RT);
+    return ((leftTrigger != _buttonPressedMap.end()) && (rightTrigger != _buttonPressedMap.end()));
+}
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: Before this lands in master, we need to use a calibration trigger that occurs less often then a double trigger press.  For example: if the MENU button and the triggers on both controllers are pressed for more then 3 seconds.

**Comment by hifi-gustavo** [comment]
Build console output for commit: bce9e9ea82a92b5f66aff1b7df8b743ff947ebe9 available at: https://gist.github.com/9dcf8524cf932ab850f2977c07ba5be8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10412-bce9e9e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10412-bce9e9e.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d688b60282047afd7e56ea66b9f2c6f42b64d4d3 available at: https://gist.github.com/9dcf8524cf932ab850f2977c07ba5be8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10412-d688b60.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10412-d688b60.dmg

**Comment by druiz17** [comment]
Test plan - 
 - perform a general smoke screen.

**Comment by hifi-gustavo** [comment]
Build console output for commit: c362ba91647cfbd7567112febca742ea4f8ffca2 available at: https://gist.github.com/9dcf8524cf932ab850f2977c07ba5be8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10412-c362ba9.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10412-c362ba9.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c362ba91647cfbd7567112febca742ea4f8ffca2 available at: https://gist.github.com/9dcf8524cf932ab850f2977c07ba5be8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10412-c362ba9.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10412-c362ba9.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 53e254152da25fc43362715c03b41e466e9238cf available at: https://gist.github.com/9dcf8524cf932ab850f2977c07ba5be8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10412-53e2541.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10412-53e2541.dmg

**Pull request merged**